### ISSUE TITLE
Docs: no-shadow and no-redeclare doc clarification/rule option request

### ISSUE THREAD

**Issue by philipbordallo** [typescript-eslint/typescript-eslint#4828]
### Suggested Changes

The rule [`no-shadow`](https://typescript-eslint.io/rules/no-shadow/) has an option [`ignoreTypeValueShadow`](https://typescript-eslint.io/rules/no-shadow#ignoretypevalueshadow), which states:

> When set to true, the rule will ignore the case when you name a type the same as a variable.
> 
> TypeScript allows types and variables to shadow one-another. This is generally safe because you cannot use variables in type locations without a typeof operator, so there's little risk of confusion.
> 
> Examples of correct code with `{ ignoreTypeValueShadow: true }`:
> 
> ```ts
> type Foo = number;
> const Foo = 1;
> 
> interface Bar {
>   prop: number;
> }
> const Bar = 'test';
> ```

The rule [`no-redeclare`](https://typescript-eslint.io/rules/no-redeclare/) on the other hand has the option [`ignoreDeclarationMerge`](https://typescript-eslint.io/rules/no-redeclare/#ignoredeclarationmerge), which states:
> **Note**: Even with this option set to true, this rule will report if you name a type and a variable the same name. **This is intentional**. Declaring a variable and a type and a variable the same is usually an accident, and it can lead to hard-to-understand code. If you have a rare case where you're intentionally naming a type the same name as a variable, use a disable comment. For example:
> ```ts
> type something = string;
> // eslint-disable-next-line @typescript-eslint/no-redeclare -- intentionally naming the variable the same as the type
> const something = 2;
> ```

With both of these set to `true` (which is the default for both options), the previous example will have the errors:
```ts
type Foo = number;
const Foo = 1;
//    ^^^ 'Foo' is already defined. ESLint(@typescript-eslint/no-redeclare)
interface Bar {
  prop: number;
}
const Bar = 'test';
//    ^^^ 'Bar' is already defined. ESLint(@typescript-eslint/no-redeclare)
```
[Playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=C4TwDgpgBAYg9nKBeKA7ArgWwEYQE4DcAUAMZyoDOwsCyUAjMUQJarD4BmAhidAEJc8UAN5EoUMHjhgAXGiy5CRAL6lyVKAKEoA5Oyo6CQA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYpwAmcSuAhlSAFwhwopkogA040eMnSYcCfEVIUEACwbUyAd1btO3EAF8gA&tsConfig=N4XyA)

The documentation for these options/rules seem to be at odds with each other, with one stating this practice is generally safe while the other implies that it's rare/bad practice.

### Questions
Should one of these be updated to have a consensus on a recommendation? And/or should [`no-redeclare`](https://typescript-eslint.io/rules/no-redeclare/) have an additional option to allow redecoration of mixed types (`const` & `type`, etc) as TypeScript does allow for it?

For example, with no rules set, TypeScript only gives errors for the first four and the last four are valid:
```ts
const redeclareConst = 1;
//    ^^^^^^^^^^^^^^ Cannot redeclare block-scoped variable 'redeclareConst'. (2451)
const redeclareConst = 1;
//    ^^^^^^^^^^^^^^ Cannot redeclare block-scoped variable 'redeclareConst'. (2451)

let redeclareLet = 1;
//  ^^^^^^^^^^^ Cannot redeclare block-scoped variable 'redeclareLet'. (2451)
let redeclareLet = 1;
//  ^^^^^^^^^^^ Cannot redeclare block-scoped variable 'redeclareLet'. (2451)

function duplicateFunction () {}
//       ^^^^^^^^^^^^^^^^^ Duplicate function implementation. (2393)
function duplicateFunction () {}
//       ^^^^^^^^^^^^^^^^^ Duplicate function implementation. (2393) 

type DuplicateType = 1;
//   ^^^^^^^^^^^^^ Duplicate identifier 'DuplicateType'. (2300)
type DuplicateType = 1;
//   ^^^^^^^^^^^^^ Duplicate identifier 'DuplicateType'. (2300)

interface MergeInterface { example: 1 }
interface MergeInterface { example: 1 }

const RedeclareMixConst = 1;
type RedeclareMixConst = 1;

let RedeclareMixLet = 1;
type RedeclareMixLet = 1;

function RedeclareMixFunction() {}
type RedeclareMixFunction = () => void;
```
[Playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=MYewdgzgLgBATgUwCYOAGwIaIMLmjAXhgEYBuAKFElkRXSwV2sJIvLQRuVU0QBlOLMu0G0eDAbCLDyAMwCuYYFACW4GEnkAHNCuAYoCAGKLlasDAAUAShgBvAL5zTq9Zp16Dxl+au3H5ORQAJ5aCDAAItq6+oYAKqHh0hQhYZHRnvGJQmwqYIZwshjA4QBK3PS0AJL5CIXF4XYwCAAeGAC2OggAXCQwTnkFRSUw5XS8yDVDDfbNbZ0cvcT9gVT4Y+KIALIqLUz4yUHZG5UIO3t4UqyBHLAnE+eSOUdp9wyPgofOSq4Wb9u7Ew-cw2exOVJlCoPQE+dREUEEAB8MAAbiAVEhSEA&rules=N4XyA&tsConfig=N4XyA)

Side note, really appreciate that y'all created a playground to see and understand what is going on.

### Affected URL(s)

https://typescript-eslint.io/rules/no-shadow/
https://typescript-eslint.io/rules/no-redeclare/

**Comment by bradzacher** [comment]
Shadowing occurs in nested scopes.
Redeclaration occurs within the same scope.

```ts
var x = 1;
var x = 2; // redeclaration

{
  var x = 3; // shadowing
}
```

The docs for `no-shadow` should probably be fixed up here. Happy to accept a PR
